### PR TITLE
Adds support for a zig based backend to zgui

### PR DESCRIPTION
**Pull request by DeanoC**
I've added a number of function to zgui, to enable a zig based backend, I'm using sokol-zig and zgui together and have it rendering with fonts etc.

These are the function and structures that need exposing, mostly so imgui can pass the draw data upto a zig backend.

### CODE REVIEW

**Comment by DeanoC** [comment]
This part is just the rendering size of things, input events coming up soon

File: libs/zgui/src/gui.zig [hunk 1]
```diff
@@ -4,6 +4,18 @@ const assert = std.debug.assert;
...
+
+pub const ImVec2 = [2]f32;
+pub const ImVec3 = [3]f32;
+pub const ImVec4 = [4]f32;
+
+pub const ImDrawIdx = u16;
+pub const ImDrawVert = struct {
+    pos: ImVec2,
+    uv: ImVec2,
+    colour: u32,
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:16]: Let's use `color` instead of `colour`.

File: libs/zgui/src/gui.zig [hunk 2]
```diff
@@ -142,6 +154,17 @@ pub const io = struct {
     pub const setDefaultFont = zguiIoSetDefaultFont;
     extern fn zguiIoSetDefaultFont(font: Font) void;
 
+    /// `pub fn getFontsTextDataAsRGBA32() return fonts pixel data and size
+    pub const getFontsTextDataAsRGBA32 = zguiIoGetFontsTexDataAsRGB32;
+    extern fn zguiIoGetFontsTexDataAsRGB32(width: *i32, height: *i32) [*c]const u32;
+
+    /// `pub fn setFontsTexID(id:TextureIdent) set the backend Id for the fonts atlas
+    pub const setFontsTexID = zguiIoSetFontsTexID;
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:162]: Let's use `setFontsTexId` name (Zig convention).

File: libs/zgui/src/gui.zig [hunk 3]
```diff
@@ -166,10 +189,24 @@ pub const io = struct {
...
+
+    /// set the frame for this frame
+    /// `pub fnsetDeltaTime(delta_time: f32) void`
+    pub const setDeltaTime = zguiIoSetDeltaTime;
+    extern fn zguiIoSetDeltaTime(delta_time: f32) void;
 };
 //--------------------------------------------------------------------------------------------------
 const Context = *opaque {};
-pub const DrawData = *opaque {};
+pub const DrawData = *extern struct {
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:200]: Struct fields should use `snake_case`.

File: libs/zgui/src/gui.zig [hunk 4]
```diff
@@ -2637,7 +2684,34 @@ extern fn zguiGetWindowDrawList() DrawList;
...
+    extern fn zguiDrawList_GetIndexBufferLength(draw_list: DrawList) c_int;
+    pub const getIndexBufferData = zguiDrawList_GetIndexBufferData;
+    extern fn zguiDrawList_GetIndexBufferData(draw_list: DrawList) [*]const ImDrawIdx;
+
+    pub const getCmdBufferLength = zguiDrawList_GetCmdBufferLength;
+    extern fn zguiDrawList_GetCmdBufferLength(draw_list: DrawList) c_int;
+    pub const getCmdBufferData = zguiDrawList_GetCmdBufferData;
+    extern fn zguiDrawList_GetCmdBufferData(draw_list: DrawList) [*]const ImDrawCmd;
+
+    pub const DrawListFlags = packed struct(u32) {
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:2703]: Struct fields should use `snake_case`.

File: libs/zgui/src/gui.zig [hunk 5]
```diff
@@ -2637,7 +2684,34 @@ extern fn zguiGetWindowDrawList() DrawList;
 extern fn zguiGetBackgroundDrawList() DrawList;
 extern fn zguiGetForegroundDrawList() DrawList;
 
-pub const DrawList = *opaque {
+pub const DrawList = *extern struct {
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:2687]: If `DrawList` has no data it should be `*opaque`.

File: libs/zgui/src/main.zig [hunk 6]
```diff
@@ -8,4 +8,4 @@ pub const version = @import("std").SemanticVersion{ .major = 0, .minor = 9, .pat
 
 pub usingnamespace @import("gui.zig");
 pub const plot = @import("plot.zig");
-pub const backend = @import("backend_glfw_wgpu.zig");
+pub const wgpu_backend = @import("backend_glfw_wgpu.zig");
```
Comment (michal-z) [inline libs/zgui/src/main.zig:11]: We can't do this, this is a breaking change.

File: libs/zgui/src/gui.zig [hunk 7]
```diff
@@ -142,6 +154,17 @@ pub const io = struct {
     pub const setDefaultFont = zguiIoSetDefaultFont;
     extern fn zguiIoSetDefaultFont(font: Font) void;
 
+    /// `pub fn getFontsTextDataAsRGBA32() return fonts pixel data and size
+    pub const getFontsTextDataAsRGBA32 = zguiIoGetFontsTexDataAsRGB32;
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:158]:
```suggestion
    pub const getFontsTextDataAsRgba32 = zguiIoGetFontsTexDataAsRGB32;
```

File: libs/zgui/src/gui.zig [hunk 8]
```diff
@@ -142,6 +154,17 @@ pub const io = struct {
...
+    extern fn zguiIoGetFontsTexDataAsRGB32(width: *i32, height: *i32) [*c]const u32;
+
+    /// `pub fn setFontsTexID(id:TextureIdent) set the backend Id for the fonts atlas
+    pub const setFontsTexID = zguiIoSetFontsTexID;
+    extern fn zguiIoSetFontsTexID(id: TextureIdent) void;
+
+    pub const getFontsTexID = zguiIoGetFontsTexID;
+    extern fn zguiIoGetFontsTexID() TextureIdent;
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:166]:
```suggestion
    extern fn zguiIoGetFontsTexId() TextureIdent;
```

File: libs/zgui/src/gui.zig [hunk 9]
```diff
@@ -142,6 +154,17 @@ pub const io = struct {
...
 
+    /// `pub fn getFontsTextDataAsRGBA32() return fonts pixel data and size
+    pub const getFontsTextDataAsRGBA32 = zguiIoGetFontsTexDataAsRGB32;
+    extern fn zguiIoGetFontsTexDataAsRGB32(width: *i32, height: *i32) [*c]const u32;
+
+    /// `pub fn setFontsTexID(id:TextureIdent) set the backend Id for the fonts atlas
+    pub const setFontsTexID = zguiIoSetFontsTexID;
+    extern fn zguiIoSetFontsTexID(id: TextureIdent) void;
+
+    pub const getFontsTexID = zguiIoGetFontsTexID;
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:165]:
```suggestion
    pub const getFontsTexId = zguiIoGetFontsTexID;
```

**Review by michal-z — CHANGES_REQUESTED** [review]
Thanks for the PR!

Overall it looks very nice, I've requested some minor changes.

Also, please do not change code convention in `zgui.cpp` (did you use auto formatter?).

**Comment by DeanoC** [comment]
All changes you asked for should be fixed. Sorry about the auto formatted zgui.cpp, didn't realise it was on by default in a new vscode install. Still getting used to zig naming convention, I think that should be all of them :)

File: libs/zgui/src/gui.zig [hunk 10]
```diff
@@ -166,10 +189,24 @@ pub const io = struct {
...
 const Context = *opaque {};
-pub const DrawData = *opaque {};
+pub const DrawData = *extern struct {
+    valid: bool,
+    cmd_lists_count: c_int,
+    total_idx_count: c_int,
+    total_vtx_count: c_int,
+    cmd_lists: [*]DrawList,
+    display_pos: @Vector(2, f32),
+    display_size: @Vector(2, f32),
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:207]:
```suggestion
    display_size: [2]f32,
```

File: libs/zgui/src/gui.zig [hunk 11]
```diff
@@ -166,10 +189,24 @@ pub const io = struct {
...
+    framebuffer_scale: @Vector(2, f32),
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:208]:
```suggestion
    framebuffer_scale: [2]f32,
```

File: libs/zgui/src/gui.zig [hunk 12]
```diff
@@ -4,6 +4,18 @@ const assert = std.debug.assert;
 //--------------------------------------------------------------------------------------------------
 pub const f32_min: f32 = 1.17549435082228750796873653722225e-38;
 pub const f32_max: f32 = 3.40282346638528859811704183484517e+38;
+
+pub const ImVec2 = [2]f32;
+pub const ImVec3 = [3]f32;
+pub const ImVec4 = [4]f32;
```
Comment (michal-z) [inline libs/zgui/src/gui.zig:10]: Let's stick to convention and use `[2]f32`, we can remove those types.

**Review by michal-z — CHANGES_REQUESTED** [review]
Thanks!

Few more minor things and also, we don't need name prefixes (`ImDrawIdx` -> `DrawIdx`, etc.).

**Comment by DeanoC** [comment]
All Vector(n,f32) replaced with [n]f32 and all Im prefixes removed. All ImVecX gone

File: libs/zgui/build.zig [hunk 13]
```diff
@@ -5,7 +5,7 @@ pub const pkg = std.build.Pkg{
     .source = .{ .path = thisDir() ++ "/src/main.zig" },
 };
 
-pub fn link(exe: *std.build.LibExeObjStep) void {
+pub fn link(exe: *std.build.LibExeObjStep, use_wgpu_backend: bool) void {
```
Comment (michal-z) [inline libs/zgui/build.zig:8]: I forgot about this, this is also a breaking change. I think we should create a new function: `linkCustomBackend`.

**Review by michal-z — CHANGES_REQUESTED** [review]
Great! Please take a look at the last issue I forgot about.

File: libs/zgui/build.zig (hunk 13 above)
Comment (DeanoC) [reply libs/zgui/build.zig]: Added a linkNoBackend function that has all the non backend parts, and link not calls that and adds the wgpu backend.
Moved the cflags to global scope as used by both functions

**Pull request merged**